### ISSUE TITLE
[bug] Fibaro Double Switch II autodiscovery

### ISSUE THREAD

**Issue by Dinth** [OpenZWave/Zwave2Mqtt#214]
**Version**

Build/Run method
- [*] Docker
- [ ] PKG
- [ ] Manually built (git clone - npm install - npm run build ) 

Zwave2Mqtt version: 2.1.1
Openzwave Version: 1.6

**Describe the bug**
First of all, im really sorry for logging another bug. I think Z2M is a great piece of software and appreciate the time and effort you are putting in this project. Please dont treat my tickets as demanding attitude, it's more like football fan supporting his club - I think that Z2M has many advantages over HA native Zwave support and I just want it to succeed. 

Ive got a problem with Fibaro Double Switch II (FGS223) HA autodiscovery.
Among 16 other devices attached to this node, Z2M detects both dimmer (incorrectly, as the device doesnt have dimming capability) and a switch (correctly) for each relay. I know that this probably can be sorted out by devices.js, but that's not the major problem here. 
The switches, while they have HASS autodiscovery JSON field showing up in Z2M panel, are not being discovered in HA. Other "devices", including dimmers detected by Z2M as part of Fibaro FGS223 are being seen in HA. 

Here's my HASS Autodisovery JSON for one of the switches:
```
{
  "type": "switch",
  "object_id": "switch_0_2",
  "discovery_payload": {
    "payload_off": false,
    "payload_on": true,
    "value_template": "{{ value }}",
    "command_topic": "zwave/nodeID_20/37/2/0/set",
    "state_topic": "zwave/nodeID_20/37/2/0",
    "device": {
      "identifiers": [
        "zwave2mqtt_nodeID_20"
      ],
      "manufacturer": "FIBARO System",
      "model": "0x1000",
      "name": "FGS223 Double Relay",
      "sw_version": "Unknown"
    },
    "name": "nodeID_20_switch_0_2",
    "unique_id": "zwave2mqtt_nodeID_20_20-37-2-0"
  },
  "discoveryTopic": "switch/nodeID_20/switch_0_2/config",
  "values": [
    "37-2-0"
  ],
  "persistent": false,
  "id": "switch_switch_0_2"
}
```

**Comment by robertsLando** [comment]
> First of all, im really sorry for logging another bug. I think Z2M is a great piece of software and appreciate the time and effort you are putting in this project. Please dont treat my tickets as demanding attitude, it's more like football fan supporting his club - I think that Z2M has many advantages over HA native Zwave support and I just want it to succeed.
 
Don't worry @Dinth I really appreciate your effort and I think every issue is a change to improve the project :smile: 

> The switches, while they have HASS autodiscovery JSON field showing up in Z2M panel, are not being discovered in HA. Other "devices", including dimmers detected by Z2M as part of Fibaro FGS223 are being seen in HA.

Check hass logs, there you will find errors thrown when they receive the discovery payload. Paste the log here

**Comment by Dinth** [comment]
BTW is there an option to force Z2M to re-send the discovery payload without restarting whole docker container? It would make troubleshooting easier.

**Comment by robertsLando** [comment]
Yes there is a rediscover button in control panel, node tab, hass devices section, on the right when you select a device from the table you see the json and on the top of it there is rediscover button

**Comment by Dinth** [comment]
Here's the log:

"2020-01-24 07:27:18 DEBUG (MainThread) [homeassistant.components.mqtt] Received message on homeassistant/switch/Living_room-Heating_1/switch/config: b'{"payload_off":false,"payload_on":true,"value_template":"{{ value }}","command_topic":"zwave/Living_room/Heating_1/37/1/0/set","state_topic":"zwave/Living_room/Heating_1/37/1/0","device":{"identifiers":["zwave2mqtt_Living_room-Heating_1"],"manufacturer":"FIBARO System","model":"0x1000","name":"FGS223 Double Relay","sw_version":"Unknown"},"name":"Living_room-Heating_1_switch","unique_id":"zwave2mqtt_Living_room-Heating_1_20-37-1-0"}'
2020-01-24 07:27:18 INFO (MainThread) [homeassistant.components.mqtt.discovery] Component has already been discovered: switch Living_room-Heating_1 switch, sending update
2020-01-24 07:27:18 INFO (MainThread) [homeassistant.components.mqtt] Got update for entity with hash: ('switch', 'Living_room-Heating_1 switch') '{'payload_off': False, 'payload_on': True, 'value_template': '{{ value }}', 'command_topic': 'zwave/Living_room/Heating_1/37/1/0/set', 'state_topic': 'zwave/Living_room/Heating_1/37/1/0', 'device': {'identifiers': ['zwave2mqtt_Living_room-Heating_1'], 'manufacturer': 'FIBARO System', 'model': '0x1000', 'name': 'FGS223 Double Relay', 'sw_version': 'Unknown'}, 'name': 'Living_room-Heating_1_switch', 'unique_id': 'zwave2mqtt_Living_room-Heating_1_20-37-1-0', 'platform': 'mqtt', 'discovery_hash': ('switch', 'Living_room-Heating_1 switch')}'
2020-01-24 07:27:18 INFO (MainThread) [homeassistant.components.mqtt] Updating component: switch.living_room_heating_1_switch
2020-01-24 07:27:27 DEBUG (MainThread) [homeassistant.components.mqtt] Received message on homeassistant/switch/Living_room-Heating_1/switch_0_2/config: b'{"payload_off":false,"payload_on":true,"value_template":"{{ value }}","command_topic":"zwave/Living_room/Heating_1/37/2/0/set","state_topic":"zwave/Living_room/Heating_1/37/2/0","device":{"identifiers":["zwave2mqtt_Living_room-Heating_1"],"manufacturer":"FIBARO System","model":"0x1000","name":"FGS223 Double Relay","sw_version":"Unknown"},"name":"Living_room-Heating_1_switch_0_2","unique_id":"zwave2mqtt_Living_room-Heating_1_20-37-2-0"}'
2020-01-24 07:27:27 INFO (MainThread) [homeassistant.components.mqtt.discovery] Component has already been discovered: switch Living_room-Heating_1 switch_0_2, sending update
2020-01-24 07:27:27 INFO (MainThread) [homeassistant.components.mqtt] Got update for entity with hash: ('switch', 'Living_room-Heating_1 switch_0_2') '{'payload_off': False, 'payload_on': True, 'value_template': '{{ value }}', 'command_topic': 'zwave/Living_room/Heating_1/37/2/0/set', 'state_topic': 'zwave/Living_room/Heating_1/37/2/0', 'device': {'identifiers': ['zwave2mqtt_Living_room-Heating_1'], 'manufacturer': 'FIBARO System', 'model': '0x1000', 'name': 'FGS223 Double Relay', 'sw_version': 'Unknown'}, 'name': 'Living_room-Heating_1_switch_0_2', 'unique_id': 'zwave2mqtt_Living_room-Heating_1_20-37-2-0', 'platform': 'mqtt', 'discovery_hash': ('switch', 'Living_room-Heating_1 switch_0_2')}'
2020-01-24 07:27:27 INFO (MainThread) [homeassistant.components.mqtt] Updating component: switch.living_room_heating_1_switch_0_2"

**Comment by robertsLando** [comment]
@Dinth No error throw. Is this an hass bug so?

**Comment by Dinth** [comment]
Potentially, I have logged https://github.com/home-assistant/home-assistant/issues/31123 this on their bug tracker

**Comment by robertsLando** [comment]
Thanks @Dinth ! As you see Hass auto discovery is relally hard, also Hass updates usually break something like the lock device in latest release :disappointed:

**Comment by Dinth** [comment]
You were absolutely correct, i have just identified why this doesnt work https://github.com/home-assistant/home-assistant/issues/31123#issuecomment-578385801

While the workaround using devices.js is possible, i will wait on what HA team will say.

**Comment by robertsLando** [comment]
@Dinth Thanks! Keep me updated

**Comment by Dinth** [comment]
@robertsLando can you have a look on this comment https://github.com/home-assistant/home-assistant/issues/31123#issuecomment-580053235. What do you think?

**Comment by robertsLando** [comment]
> It seems strange that the false/true in the discovery message are not strings. My understanding is that the payload values in MQTT are strings, not booleans. if true, this would be a bug in zwave2mqtt.

That has no sense. The payload is parsed into into json and json supports Boolean types.

Now by checking your discovery payload:

```
"discovery_payload": {
    "payload_off": false,
    "payload_on": true,
    "value_template": "{{ value_json.value }}", // <--- this was wrong
    "command_topic": "zwave/nodeID_20/37/2/0/set",
    "state_topic": "zwave/nodeID_20/37/2/0",
    "device": {
      "identifiers": [
        "zwave2mqtt_nodeID_20"
      ],
      "manufacturer": "FIBARO System",
      "model": "0x1000",
      "name": "FGS223 Double Relay",
      "sw_version": "Unknown"
    },
    "name": "nodeID_20_switch_0_2",
    "unique_id": "zwave2mqtt_nodeID_20_20-37-2-0"
  },
```

**Comment by robertsLando** [comment]
I have submitted a fix for this

**Comment by Dinth** [comment]
Hi @robertsLando 

I havent checked the nodejs version yet, due to issues at work, but i changed the HASS discovery template to this:
```
{
  "type": "switch",
  "object_id": "switch",
  "discovery_payload": {
    "payload_off": false,
    "payload_on": true,
    "value_template": "{{ value | bool }}",
    "command_topic": "zwave/Living_room/Radiator_1/37/1/0/set",
    "state_topic": "zwave/Living_room/Radiator_1/37/1/0",
    "device": {
      "identifiers": [
        "zwave2mqtt_Living_room-Radiator_1"
      ],
      "manufacturer": "FIBARO System",
      "model": "0x1000",
      "name": "FGS223 Double Relay",
      "sw_version": "Unknown"
    },
    "name": "Living_room-Radiator_1_switch",
    "unique_id": "zwave2mqtt_Living_room-Radiator_1_20-37-1-0"
  },
  "discoveryTopic": "switch/Living_room-Radiator_1/switch/config",
  "values": [
    "37-1-0"
  ],
  "persistent": false,
  "id": "switch_switch"
}
```
and now HA is throwing out this error:

> voluptuous.error.MultipleInvalid: invalid template (TemplateAssertionError: no filter named 'bool') for dictionary value @ data['value_template']
> 2020-01-30 19:49:10 ERROR (MainThread) [homeassistant.util.logging] Exception in async_discover when dispatching 'mqtt_discovery_new_switch_mqtt': ({'payload_off': False, 'payload_on': True, 'value_template': '{{ value | bool }}', 'command_topic': 'zwave/Living_room/Radiator_1/37/2/0/set', 'state_topic': 'zwave/Living_room/Radiator_1/37/2/0', 'device': {'identifiers': ['zwave2mqtt_Living_room-Radiator_1'], 'manufacturer': 'FIBARO System', 'model': '0x1000', 'name': 'FGS223 Double Relay', 'sw_version': 'Unknown'}, 'name': 'Living_room-Radiator_1_switch_0_2', 'unique_id': 'zwave2mqtt_Living_room-Radiator_1_20-37-2-0', 'platform': 'mqtt'},)
> Traceback (most recent call last):
>   File "/usr/src/homeassistant/homeassistant/components/mqtt/switch.py", line 81, in async_discover
>     config = PLATFORM_SCHEMA(discovery_payload)
>   File "/usr/local/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 272, in __call__
>     return self._compiled([], data)
>   File "/usr/local/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 594, in validate_dict
>     return base_validate(path, iteritems(data), out)
>   File "/usr/local/lib/python3.7/site-packages/voluptuous/schema_builder.py", line 432, in validate_mapping
>     raise er.MultipleInvalid(errors)
> voluptuous.error.MultipleInvalid: invalid template (TemplateAssertionError: no filter named 'bool') for dictionary value @ data['value_template']

**Comment by robertsLando** [comment]
@Dinth Sorry my fault, jinja doesn't support booloean casts, try with:

`"value_template": "{{ value == 'true' }}"`

**Comment by Dinth** [comment]
Done this, the errors are gone now and i think that the switch works now. Need to reset whole my system, including mosquitto to double check but it looks promising.

**Comment by Dinth** [comment]
Yep. im not sure what == 'true' does but it seems to work :D

**Comment by robertsLando** [comment]
@Dinth Already done, thanks! :)
